[Static Analyzer CI] Hide unexpected results link if there are no unexpected results

briannafan opened the pull request:
#### 1c1374263cc64cdf8a8b080703e7d6b27e615810
<pre>
[Static Analyzer CI] Hide unexpected results link if there are no unexpected results
<a href="https://bugs.webkit.org/show_bug.cgi?id=279289">https://bugs.webkit.org/show_bug.cgi?id=279289</a>
<a href="https://rdar.apple.com/129720143">rdar://129720143</a>

Reviewed by Aakash Jain.

Add result links to DisplaySmartPointerResults and hide unexpected when there are none.

* Tools/CISupport/build-webkit-org/steps.py:
(DisplaySmartPointerResults): Rename DisplayUnexpectedResults to DisplaySmartPointerResults.
(DisplaySmartPointerResults.run): Add result links to DisplaySmartPointerResults instead of ExtractStaticAnalyzerTestResults.
(DisplaySmartPointerResults.getFilesPerProject): Improve logging.
(DisplaySmartPointerResults.getResultSummary): Display &apos;No unexpected results&apos; upon success.
(DisplaySmartPointerResults.resultDirectoryURL): Added.
(DisplaySmartPointerResults.resultDownloadURL): Added.
(ExtractStaticAnalyzerTestResults):
(ExtractStaticAnalyzerTestResults.addCustomURLs): Pass.
(ExtractStaticAnalyzerTestResults.run): Deleted.
(ExtractStaticAnalyzerTestResults.getLastBuildStepByName): Deleted.
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/283439@main">https://commits.webkit.org/283439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/72004f3eaf39548d29349c9a2d23436a5866c5af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45655 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70314 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16892 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68400 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17173 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53170 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11774 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69349 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57384 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38772 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15768 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15109 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72017 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10237 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14500 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/65962 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10269 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60794 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8451 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2077 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10043 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43722 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42283 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1724,10 +1724,10 @@ def getResultSummary(self):
-    name = 'compare-static-analyzer-results'
-    description = ['comparing static analyzer results']
-    descriptionDone = ['compared static analyzer results']
+class FindUnexpectedStaticAnalyzerResults(shell.ShellCommandNewStyle):
```
This rename seems like a unrelated change, better to do it in separate PR. Mixing multiple separate logical things together makes it harder to review.

briannafan replied:
Sorry about that! I'll be sure to do this in the future too.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1776,39 +1774,58 @@ def getResultSummary(self):
                     log_content += f'\n\n=> {checker}\n\n'
-                    log_content += '\n'.join(files)
+                    log_content += '\n'.join((sorted(files)))
+                    log_content += '\n\n'
```
Nit: can use f-string to combine all these three lines into one.

briannafan replied:
I made the expression more compact, but I also learned that '\n' can't go in f-strings so that expression is on a separate line.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1776,39 +1774,58 @@ def getResultSummary(self):
     def getResultSummary(self):
-        if self.results != SUCCESS:
-            return super().getResultSummary()
+        if self.results == SUCCESS:
```
Please verify if the logic is expected after this inversion. Also verify the case for WARNINGS.

briannafan replied:
Yes, this is correct. When there are no unexpected results (SUCCESS), it's clearer to show "No unexpected results" than "Unexpected failing files: 0 Unexpected passing files: 0". I added different summary for warnings.

aj062 reviewed (commented):
> Hide unexpected results link if there are no unexpected results

Now, people might complain about missing results link 😆 Make sure that is clear to people.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1776,39 +1774,59 @@ def getResultSummary(self):
+        if not is_log:
+            yield self.addToLog('stdio', 'No unexpected results.\n')
+
+        self.addURL("View full static analyzer results", self.resultDirectoryURL() + SCAN_BUILD_OUTPUT_DIR + "/results.html")
```
Do you  want to keep these url adding logic in addCustomURLs()?

briannafan replied:
If you mean keep it under `ExtractTestResults`, unfortunately that doesn't work since `DisplaySmartPointerResults` is run after `ExtractTestResults`.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1776,39 +1774,59 @@ def getResultSummary(self):
     def run(self):
-        result_directory = f"public_html/results/{self.getProperty('buildername')}/{self.getProperty('archive_revision')} ({self.getProperty('buildnumber')})"
-        unexpected_results_json = os.path.join(result_directory, SCAN_BUILD_OUTPUT_DIR, 'unexpected_results.json')
+        self.zipFile = f"public_html/results/{self.getProperty('buildername')}/{self.getProperty('archive_revision')} ({self.getProperty('buildnumber')}).zip"
```
This path looks similar to  self.resultDirectory. Do you want to just re-use the path from that variable to avoid duplication? something like:
`self.zipFile = f"{self.resultDirectory}.zip"`

briannafan committed:
[Static Analyzer CI] Hide unexpected results link if there are no unexpected results
https://bugs.webkit.org/show_bug.cgi?id=279289
rdar://129720143

Reviewed by Aakash Jain.

Add result links to DisplaySmartPointerResults and hide unexpected when there are none.

* Tools/CISupport/build-webkit-org/steps.py:
(DisplaySmartPointerResults): Rename DisplayUnexpectedResults to DisplaySmartPointerResults.
(DisplaySmartPointerResults.run): Add result links to DisplaySmartPointerResults instead of ExtractStaticAnalyzerTestResults.
(DisplaySmartPointerResults.getFilesPerProject): Improve logging.
(DisplaySmartPointerResults.getResultSummary): Display 'No unexpected results' upon success.
(DisplaySmartPointerResults.resultDirectoryURL): Added.
(DisplaySmartPointerResults.resultDownloadURL): Added.
(ExtractStaticAnalyzerTestResults):
(ExtractStaticAnalyzerTestResults.addCustomURLs): Pass.
(ExtractStaticAnalyzerTestResults.run): Deleted.
(ExtractStaticAnalyzerTestResults.getLastBuildStepByName): Deleted.
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: https://commits.webkit.org/283439@main
